aws/iam.go

```diff
@@ -0,0 +1,417 @@
+		}
+
+		logging.Logger.Infof("Deleted Virtual MFA Device with ID %s from user %s", aws.StringValue(serialNumber), aws.StringValue(userName))
+	}
+
+	return nil
+}
+
+func deleteUser(svc *iam.IAM, userName *string) error {
+	_, err := svc.DeleteUser(&iam.DeleteUserInput{
+		UserName: userName,
+	})
+	if err != nil {
+		return errors.WithStackTrace(err)
+	}
+
+	return nil
+}
+
+// Delete all IAM Users
+func nukeAllIamUsers(session *session.Session, userNames []*string) error {
+	if len(userNames) == 0 {
+		logging.Logger.Info("No IAM Users to nuke")
+		return nil
+	}
+
+	logging.Logger.Info("Deleting all IAM Users")
+
+	deletedUsers := 0
+	svc := iam.New(session)
+
+	multiErr := new(multierror.Error)
+	for _, userName := range userNames {
+		var err error
+
+		// TODO: Add CLI option to delete the Policy as policies exist independently of the user
+		// Detach User Policies
+		err = detachUserPolicies(svc, userName)
+		if err != nil {
+			logging.Logger.Errorf("[Failed] %s", err)
+			multierror.Append(multiErr, err)
+			continue
+		}
+
+		// Delete inline user policies
+		err = deleteInlineUserPolicies(svc, userName)
+		if err != nil {
+			logging.Logger.Errorf("[Failed] %s", err)
+			multierror.Append(multiErr, err)
+			continue
+		}
+
+		// TODO: Add CLI option to delete groups as groups exist independently of the user
+		// Remove user from groups
+		err = removeUserFromGroups(svc, userName)
+		if err != nil {
+			logging.Logger.Errorf("[Failed] %s", err)
+			multierror.Append(multiErr, err)
+			continue
+		}
+
+		// Delete Login Profile
+		err = deleteLoginProfile(svc, userName)
+		if err != nil {
+			logging.Logger.Errorf("[Failed] %s", err)
+			multierror.Append(multiErr, err)
+			continue
+		}
+
+		// Delete Access Keys
+		err = deleteAccessKeys(svc, userName)
+		if err != nil {
+			logging.Logger.Errorf("[Failed] %s", err)
+			multierror.Append(multiErr, err)
+			continue
+		}
+
+		// Delete Signing Certificate
+		err = deleteSigningCertificate(svc, userName)
+		if err != nil {
+			logging.Logger.Errorf("[Failed] %s", err)
+			multierror.Append(multiErr, err)
+			continue
+		}
+
+		// Delete Public SSH Key
+		err = deleteSSHPublicKeys(svc, userName)
+		if err != nil {
+			logging.Logger.Errorf("[Failed] %s", err)
+			multierror.Append(multiErr, err)
+			continue
+		}
+
+		// Delete Service Specific Credentials (codecommit and Amazon Keyspaces)
+		err = deleteServiceSpecificCredentials(svc, userName)
+		if err != nil {
+			logging.Logger.Errorf("[Failed] %s", err)
+			multierror.Append(multiErr, err)
+			continue
+		}
+
+		// Delete MFA Devices
+		err = deleteMFADevices(svc, userName)
+		if err != nil {
+			logging.Logger.Errorf("[Failed] %s", err)
+			multierror.Append(multiErr, err)
+			continue
+		}
+
+		// Delete the user
+		// NOTE: The actual user deletion should always be the last step.
+		// This way we can guarantee that it will fail if we forgot to delete/detach an item
+		err = deleteUser(svc, userName)
```

Comment:
NIT: This inner block should be a function. It would greatly simplify this logic, as you can return `err` to shortcircuit and only have to remember to add to multiErr in one place.

Reply:
You mean moving all `detach*/delete*` functions to a new one that simply returns an `error` (as soon as there is one) so we can have something like this?
```go
if err := deleteAll(...); err != nil {
  logging.Logger.Errorf("[Failed] %s", err)
  multierror.Append(multiErr, err)
  continue
}
```

Reply:
Yes, although you won't need to do the `continue` at that point and you can use the if block immediately following this line I commented on.

Reply:
By the way I just did the following and it worked as expected :smile: 
```go
func nukeUser(svc *iam.IAM, userName *string) error {
	functions := []func(svc *iam.IAM, userName *string) error{
		detachUserPolicies, // TODO: Add CLI option to delete the Policy as policies exist independently of the user
		deleteInlineUserPolicies,
		removeUserFromGroups, // TODO: Add CLI option to delete groups as groups exist independently of the user
		deleteLoginProfile,
		deleteAccessKeys,
		deleteSigningCertificate,
		deleteSSHPublicKeys,
		deleteServiceSpecificCredentials,
		deleteMFADevices,
		deleteUser, // NOTE: The actual user deletion should always be the last step.
		// This way we can guarantee that it will fail if we forgot to delete/detach an item
	}

	for _, fn := range functions {
		err := fn(svc, userName)
		if err != nil {
			return err
		}
	}

	return nil
}

// Delete all IAM Users
func nukeAllIamUsers(session *session.Session, userNames []*string) error {
	if len(userNames) == 0 {
		logging.Logger.Info("No IAM Users to nuke")
		return nil
	}

	logging.Logger.Info("Deleting all IAM Users")

	deletedUsers := 0
	svc := iam.New(session)
	multiErr := new(multierror.Error)

	for _, userName := range userNames {
		err := nukeUser(svc, userName)
		if err != nil {
			logging.Logger.Errorf("[Failed] %s", err)
			multierror.Append(multiErr, err)
		} else {
			deletedUsers++
			logging.Logger.Infof("Deleted IAM User: %s", *userName)
		}
	}

	logging.Logger.Infof("[OK] %d IAM User(s) terminated", deletedUsers)
	return multiErr.ErrorOrNil()
}
```

Was it something like this? :thinking:

Reply:
Yup this is what I had in mind!